draft-ietf-quic-transport.md

```diff
@@ -1473,27 +1473,30 @@ contains a FIN flag or when either peer sends a RST_STREAM frame.
 
 The "closed" state is the terminal state.
 
-A final offset is present in both a frame bearing a FIN flag and in a RST_STREAM
-frame.  Upon sending either of these frames for a stream, the endpoint MUST NOT
-send a STREAM frame carrying data beyond the final offset.
+A final offset is present in both a STREAM frame bearing a FIN flag and in a
+RST_STREAM frame.  Upon sending either of these frames for a stream, the
+endpoint MUST NOT send a STREAM frame carrying data beyond the final offset.
 
-An endpoint that receives any frame for this stream after receiving either a FIN
-flag and all stream data preceding it, or a RST_STREAM frame, MUST quietly
-discard the frame, with one exception.  If a STREAM frame carrying data beyond
-the received final offset is received, the endpoint MUST close the connection
-with a QUIC_STREAM_DATA_AFTER_TERMINATION error ({{error-handling}}).
+An endpoint that receives a FIN flag or a RST_STREAM frame knows the final
+offset for this stream. If a STREAM frame carrying data beyond the received
+final offset is received, the endpoint MUST close the connection with a
+QUIC_STREAM_DATA_AFTER_TERMINATION error ({{error-handling}}).
+
+An endpoint MUST close the connection with a QUIC_STREAM_DATA_AFTER_TERMINATION
+error if it receives a RST_STREAM frame or a FIN flag with a lower final offset
+than the highest data offset it already received on that stream.
```

Comment:
This can be simplified greatly: 

If a STREAM frame carries data with an offset higher than the final offset, {{ERROR}}.  If a RST_STREAM or a STREAM frame marked with a FIN flag indicates a different final offset, {{ERROR}}.